substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java

```diff
@@ -342,11 +344,27 @@ public ProviderList cleanUnregisteredProviders(ProviderList providerList) {
             filteredProviders.removeIf(this::shouldRemoveProvider);
             if (filteredProviderList == null || !filteredProviderList.providers().equals(filteredProviders)) {
                 filteredProviderList = ProviderList.newList(filteredProviders.toArray(new Provider[0]));
+                removedProviders = new ArrayList<>(providerList.providers());
+                removedProviders.removeIf(provider -> !shouldRemoveProvider(provider));
             }
         }
         return filteredProviderList;
     }
 
+    private void traceRemovedProviders() {
+        if (removedProviders.isEmpty()) {
```

Comment:
```suggestion
        if (removedProviders == null || removedProviders.isEmpty()) {
```

Reply:
Nice catch! :)

Reply:
Well, it was the CI https://github.com/oracle/graal/pull/3473/checks?check_run_id=2819581300#step:9:788 not me :)

Reply:
Ah, I haven't run the gates on this internally yet, didn't check these :)
The reason a NPE could happen here - when security isn't used, analysis wont see the `Providers.providerList` field. This means the field value recomputer will never run and thus never clear any providers, resulting in a NPE when we access the `removedProviders` list